contracts/external/cw-verifier-middleware/src/verify.rs

```diff
@@ -0,0 +1,164 @@
+use crate::{
+    error::ContractError,
+    msg::WrappedMessage,
+    state::{CONTRACT_ADDRESS, NONCES},
+};
+use bech32::{ToBase32, Variant};
+use cosmwasm_schema::{schemars::_serde_json::json, serde::de::DeserializeOwned};
+use cosmwasm_std::{
+    from_slice, to_binary, Addr, Api, DepsMut, Env, MessageInfo, StdError, Storage, Uint128,
+};
+
+use ripemd::Ripemd160;
+use sha2::{Digest, Sha256};
+
+const UNCOMPRESSED_HEX_PK_LEN: usize = 130;
+const COMPRESSED_HEX_PK_LEN: usize = 66;
+
+pub fn verify<T>(
+    api: &dyn Api,
+    storage: &mut dyn Storage,
+    env: &Env,
+    info: MessageInfo,
+    wrapped_msg: WrappedMessage,
+) -> Result<(T, MessageInfo), ContractError>
+where
+    T: DeserializeOwned,
+{
+    let payload = wrapped_msg.payload;
+
+    let signer_addr = pk_to_addr(
+        api,
+        wrapped_msg.public_key.to_hex(), // to_hex ensures that the public key has the expected number of bytes
+        &payload.bech32_prefix,
+    )?;
+
+    let payload_ser = serde_json::to_string(&payload)?;
+
+    // Convert message to signDoc format
+    let sign_doc = get_sign_doc(&signer_addr.as_str(), &payload_ser, &payload.chain_id)?;
+
+    // Serialize the payload
+    let msg_ser = to_binary(&sign_doc)?;
+
+    // Hash the serialized payload using SHA-256
+    let msg_hash = Sha256::digest(&msg_ser);
+
+    // Verify the signature
+    let sig_valid = api.secp256k1_verify(
+        msg_hash.as_slice(),
+        &wrapped_msg.signature,
+        wrapped_msg.public_key.as_slice(),
+    )?;
+
+    if !sig_valid {
+        return Err(ContractError::SignatureInvalid {});
+    }
+
+    // Validate that the message has not expired
+    if let Some(expiration) = payload.expiration {
+        if expiration.is_expired(&env.block) {
+            return Err(ContractError::MessageExpired {});
+        }
+    }
+
+    let validated_contract_addr = api.addr_validate(&payload.contract_address)?;
+    let pk = wrapped_msg.public_key.to_hex();
+    let nonce_key = (
+        pk.as_str(),
+        &validated_contract_addr,
+        payload.contract_version.as_str(),
+    );
+
+    // Validate that the message has the correct nonce
+    let nonce = NONCES
+        .may_load(storage, nonce_key)?
+        .unwrap_or(Uint128::from(0u128));
+
+    if payload.nonce != nonce {
+        return Err(ContractError::InvalidNonce {});
+    }
+
+    // Increment nonce
+    NONCES.update(storage, nonce_key, |nonce: Option<Uint128>| {
+        nonce
+            .unwrap_or(Uint128::from(0u128))
```

Comment:
i'm tempted to say `unwrap_or_default()`, or `unwrap_or(Uint128::zero())`, but i like the brutalist simplicity of this. this code has incredible aesthetic.